### PR TITLE
[WIP] X509: Refactor X509_PUBKEY processing to include provider side keys

### LINKED ISSUE #13893: Public keys decoded from an X509 certificate are legacy keys
When we decode an X509 certificate, the public key contained within it is decoded by the internal function `x509_pubkey_decode`:

https://github.com/openssl/openssl/blob/47b784a41b729d5df9ad47c99355db2f2026a709/crypto/x509/x_pubkey.c#L181-L214

This firstly takes the algorithm and uses `OBJ_obj2nid` to find a nid for it. This is the first problem. What if the algorithm comes from a provider and we don't have a nid for that algorithm? Once we have the nid we call `EVP_PKEY_set_type` which has the effect of setting the ameth on the EVP_PKEY. Subsequently we then call `ameth->pub_decode` to decode it - which obviously gives us a legacy key.

Instead we somehow need to find a decoder from a provider that is associated with the algorithm OID and use that (perhaps using "capabilities"?). Once we have the decoder we can use that to create a provider side key.

### PR DESCRIPTION
**Pull request by levitte**
When a SubjectPublicKeyInfo (SPKI) is decoded into an X509_PUBKEY
structure, the corresponding EVP_PKEY is automatically added as well.
This used to only support our built-in keytypes, and only in legacy
form.

This is now refactored to try to decode the whole SPKI into an
EVP_PKEY using OSSL_DECODER, and cache that in the X509_PUBKEY very
early.  Only when that fails, the existing decoder callback will try
to create a legacy EVP_PKEY.

For purposes where it's not desirable to do the provider side SPKI
decoding, there is now a function d2i_PUBKEY_legacy().  This is used
by all the type specific d2i_<TYPE>_PUBKEY(), as well as in out
provider DER->key decoder, to avoid running into recursion problems.

Fixes #13893

-----

This includes refactoring the DER-\>key decoder to be a bit more flexible with how it uses keytype specific d2i functions and adds the possibility to use d2i_*TYPE*_PUBKEY.

This also includes the addition of SM2 encoders and decoders...

### CODE REVIEW

**Comment by levitte** [comment]
Something to be noted is the inherent problem with the early caching of an EVP_PKEY in the X509_PUBKEY, when someone else wants an easy way to decode the DER, as part of a decoder implementation for a keytype we know nothing about.  It becomes a recursion problem.

The solution I could come up with was to have a `d2i_PUBKEY_legacy()`, which is currently internal.  However, that doesn't help anyone making an external provider for a keytype that's unknown to us, and that merely wants to decode the SubjectPublicKeyInfo into its parts, but no further processing (i.e. no automatic caching of an EVP_PKEY).

**Comment by levitte** [comment]
A slightly different solution could have been to let `x509_public_decode()` handle the decoding via OSSL_DECODER.  The only issue there is that keytypes like RSA-PSS would be a bit problematic, as its parameters aren't included in the pubkey BIT STRING itself, but rather in the AlgorithmID (X509_ALGOR) that sits beside it.

This could be resolved by decoding the X509_ALGOR `parameters` field and the X509_PUBKEY `pubkey` fields separately, which is possible with the help of d2i_*TYPE*PublicKey() and d2i_*TYPE*params(), assign what we get to EVP_PKEYs, and use EVP_PKEY_copy_parameters() to copy the parameters extracted from the X509_ALGOR onto the EVP_PKEY that contains the key itself...  the only issue there being we don't have  d2i_*TYPE*PublicKey() and d2i_*TYPE*params() for all our legacy keytypes.

**Comment by mattcaswell** [comment]
Hmm, this solution doesn't seem to work for me. If I run this code:

````
        x509 = PEM_read_bio_X509(bio, NULL, NULL, NULL);
        assert(x509 != NULL);
        pkey = X509_get_pubkey(x509);

        assert(pkey != NULL);
        assert(pkey->keymgmt != NULL);
````

I get:
````
readx509-2: readx509-2.c:24: main: Assertion `pkey->keymgmt != NULL' failed.
Aborted (core dumped)
````

**Comment by levitte** [comment]
Yeah, something is off. See the CI failures

**Comment by levitte** [comment]
What key type was there in that certificate?  I just found a problem where RSA and RSA-PSS gets mixed up...

**Comment by mattcaswell** [comment]
RSA

**Comment by levitte** [comment]
A number of fixups added.  @mattcaswell, would you mind trying again?

File: crypto/x509/x_pubkey.c [hunk 1]
```diff
@@ -33,6 +34,14 @@ struct X509_pubkey_st {
     /* extra data for the callback, used by d2i_PUBKEY_ex */
     OSSL_LIB_CTX *libctx;
     char *propq;
+
+    /*
+     * d2i_X509_PUBKEY() does double parsing, first with OSSL_DECODER,
+     * and the with ASN1_item_d2i().  It leaves the result of OSSL_DECODER
```
Comment (slontis) [inline crypto/x509/x_pubkey.c:40]: then

**Comment by mattcaswell** [comment]
> A number of fixups added. @mattcaswell, would you mind trying again?

Same problem.

File: crypto/x509/x_pubkey.c (hunk 1 above)
Comment (levitte) [reply crypto/x509/x_pubkey.c:40]: Fixed

**Comment by levitte** [comment]
But hold on, @mattcaswell...  is there point that the EVP_PKEY *must* be provider backed?  I thought the primary problem was that it now *may* be provider backed.  In other words, provider backed EVP_PKEYs were *completely inaccessible* when decoding an X509_PUBKEY before this change, i.e. other algos than our built in ones were impossible.

**Comment by t8m** [comment]
> But hold on, @mattcaswell... is there point that the EVP_PKEY _must_ be provider backed? I thought the primary problem was that it now _may_ be provider backed. In other words, provider backed EVP_PKEYs were _completely inaccessible_ when decoding an X509_PUBKEY before this change, i.e. other algos than our built in ones were impossible.

If it is not provider backed, will it be automatically imported to provider if you do for example signature verification of a certificate with the X509_PUBKEY? IMO if that's the case, I think we can keep that.

**Comment by mattcaswell** [comment]
It remains unclear to me why we would do things differently for key types that happen to have a legacy implementation vs for those that don't. If we've got the codepaths to handle provider based keys - why wouldn't we always do that if there is a provider available to do it?

**Comment by levitte** [comment]
> If it is not provider backed, will it be automatically imported to provider if you do for example signature verification of a certificate with the X509_PUBKEY? IMO if that's the case, I think we can keep that.

Yes, this was implemented early.  The keys are exported to the provider that is to do the signature work if that provider doesn't match the key's provider (or that the key is legacy).  This is among the earliest things we implemented.

**Comment by levitte** [comment]
> It remains unclear to me why we would do things differently for key types that happen to have a legacy implementation vs for those that don't. If we've got the codepaths to handle provider based keys - why wouldn't we always do that if there is a provider available to do it?

Generally speaking, priorities. Do operations work with legacy EVP_PKEYs as well as with provider native EVP_PKEYs?  In that case, it shouldn't matter whether X509_get_pubkey() one or the other.
The main problem here was that if the algorithm given by the AlgorithmID *isn't known by OpenSSL*, then `X509_get_pubkey()` would obviously return NULL...  was it not?

**Comment by mattcaswell** [comment]
> Generally speaking, priorities (we've all enough on our plates as it is). Do operations work with legacy EVP_PKEYs as well as with provider native EVP_PKEYs? In that case, it shouldn't matter whether X509_get_pubkey() one or the other.
The main problem here was that if the algorithm given by the AlgorithmID isn't known by OpenSSL, then X509_get_pubkey() would obviously return NULL... was it not?

That remains the primary issue - but I don't understand why it is any harder to not just do the same in all cases. Surely this is easier? Why does it behave differently in these cases?

**Comment by levitte** [comment]
> That remains the primary issue - but I don't understand why it is any harder to not just do the same in all cases. Surely this is easier? Why does it behave differently in these cases?

I'm frankly not sure what you mean with "different"

**Comment by levitte** [comment]
I've tested the code that @mattcaswell used, and it turns out that the solution here isn't right.  I'll have to revisit this

**Comment by levitte** [comment]
> I'll have to revisit this

I knew there had to be a way:

https://github.com/openssl/openssl/blob/a2a5506b9329b978a2a5b11a518b9789446ad310/include/openssl/asn1t.h.in#L63-L66
https://github.com/openssl/openssl/blob/a2a5506b9329b978a2a5b11a518b9789446ad310/include/openssl/asn1t.h.in#L609-L610
https://github.com/openssl/openssl/blob/a2a5506b9329b978a2a5b11a518b9789446ad310/include/openssl/asn1t.h.in#L652-L660

Using that form allows a programmatic "take-over" of the processing of an item.  X509_NAME is processed this way, and it seems like the solution for our problem here is to do the same for X509_PUBKEY...

File: providers/implementations/encode_decode/decode_der2key.c [hunk 2]
```diff
@@ -277,59 +288,66 @@ static int der2key_decode(void *vctx, OSSL_CORE_BIO *cin, int selection,
...
-        if ((selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0) {
-            key = ctx->desc->d2i_private_key(NULL, &derp, der_len);
-            if (key == NULL && orig_selection != 0)
-                goto end;
-        }
-        if (key == NULL
-            && (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0) {
+        goto next;
+
+    /* We try the typs specific functions first, if available */
```
Comment (slontis) [inline providers/implementations/encode_decode/decode_der2key.c:293]: type

File: crypto/x509/x_pubkey.c [hunk 3]
```diff
@@ -91,9 +105,52 @@ ASN1_SEQUENCE_cb(X509_PUBKEY, pubkey_cb) = {
...
+    EVP_PKEY *pkey = NULL;
+    X509_PUBKEY *tmp_pk = NULL;
+    const unsigned char *p = *in;
+    size_t size = (size_t)len;
+    OSSL_DECODER_CTX *dctx = NULL;
+
+    if ((dctx = OSSL_DECODER_CTX_new_by_EVP_PKEY(&pkey,
+                                                 "DER", "SubjectPublicKeyInfo",
+                                                 NULL, EVP_PKEY_PUBLIC_KEY,
+                                                 NULL, NULL)) != NULL
```
Comment (slontis) [inline crypto/x509/x_pubkey.c:123]: Why is there no libctx, propq here?

File: providers/implementations/encode_decode/decode_der2key.c [hunk 4]
```diff
@@ -456,8 +489,18 @@ static void dsa_adjust(void *key, struct der2key_ctx_st *ctx)
 # define ec_d2i_private_key             (d2i_of_void *)d2i_ECPrivateKey
 # define ec_d2i_public_key              NULL
 # define ec_d2i_key_params              (d2i_of_void *)d2i_ECParameters
+# define ec_d2i_PUBKEY                  (d2i_of_void *)d2i_EC_PUBKEY
 # define ec_free                        (free_key_fn *)EC_KEY_free
 
+static int ec_check(void *key, struct der2key_ctx_st *ctx)
+{
+    /* We're trying to be clever by comparing two truths */
```
Comment (slontis) [inline providers/implementations/encode_decode/decode_der2key.c:497]: This comment seems to be a riddle..

Comment (levitte) [reply providers/implementations/encode_decode/decode_der2key.c:497]: Not much, read the code...

**Comment by levitte** [comment]
I've made this WIP because of the needed rework

**Comment by levitte** [comment]
I'm closing in favor of a new PR where I've started over

**Comment by levitte** [comment]
See #13281

**Comment by slontis** [comment]
> See #13281

You mean  #14281

**Comment by levitte** [comment]
Ah, oops, yes

**Pull request closed without merging**